### ISSUE TITLE
HTML mails have wrong Content-Type

### ISSUE THREAD

**Issue by ghmail** [roundcube/roundcubemail#7438]
The attached mail was written with the built-in message composer in mode HTML and stored as draft. Being a multipart mail the Content-Type "text/plain" is wrong. Also the boundary delimiter is not defined in the header.

[Roundcube-HTML-mail.txt](https://github.com/roundcube/roundcubemail/files/4783051/Roundcube-HTML-mail.txt)

**Comment by alecpl** [comment]
There was a report like this in #7386, but without a solution. Please, describe your environment/versions. Disable all plugins. Check Mail_mime version in use.

**Comment by ghmail** [comment]
Hi Aleksander,
I checked with all plugins disabled: same result.
My environment is as follows:
System: Linux, Kernel: 4.12.14-lp151.28.48-default
PHP: 7.2.5
php7-pear-Mail_Mime: 1.10.1
Apache: 2.4.33
Dovecot: 2.3.10

Received HTML mails are correctly shown. The error.log shows no entries.

**Comment by alecpl** [comment]
And Roundcube version? Is it a distribution package or vanilla package? I'd check with most recent Mail_mime version.

**Comment by ghmail** [comment]
Roundcube 1.4.6.
I just read #7386. I have a suspiciously similar environment (openSUSE 15.1). I have installed the Mail_mime version from the distribution repo but I have also a Mail_mime version in the vendor subdir of roundcube.
If I could support you, e.g. to patch a file with some extra log output for debugging, you're welcome.

**Comment by alecpl** [comment]
How about uninstalling the Mail_mime distribution package? Also, check what version is the lib in the vendor folder (it will be in composer.lock file, if you have one).

**Comment by alecpl** [comment]
Also, running `composer update` might help.

**Comment by ghmail** [comment]
Uninstalled system-wide Mail_mime.
`composer update` gives me the following output:
```
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 3 installs, 0 updates, 0 removals
  - Installing pear/pear_exception (v1.0.1): Loading from cache
  - Installing pear/console_getopt (v1.4.3): Loading from cache
  - Installing pear/pear-core-minimal (v1.10.10): Loading from cache
Writing lock file
Generating autoload files
```
Same result (after restart apache, of course).

**Comment by alecpl** [comment]
Does that mean you're now using Mail_mime 1.10.8? Could you verify with composer.lock?

**Comment by ghmail** [comment]
My composer.lock in _roundcube/vendor/pear/mail_mime_ contains no version information about Mail_mime. 
Just the 3 version numbers as in my last comment.
[composer.lock.txt](https://github.com/roundcube/roundcubemail/files/4785359/composer.lock.txt)

**Comment by alecpl** [comment]
So, that's fishy. What do you have in the composer.json file?

**Comment by ghmail** [comment]
[composer.json.txt](https://github.com/roundcube/roundcubemail/files/4785401/composer.json.txt)

**Comment by alecpl** [comment]
I meant the composer.json in the roundcube directory root, and the same about `composer update`.

**Comment by ghmail** [comment]
OK, I see.
`composer update` gives me "nothing to install or update"
_composer.lock_ contains version "1.10.8" for pear/mail_mime.

**Comment by alecpl** [comment]
So, it looks like we're on the same page now. However, I'm unable to reproduce.

Make sure there's no other mimePart.php file in system-wide PHP include dir. Also, did you try with a different skin? What are the precise steps you do in the mail compose page?

**Comment by alecpl** [comment]
I'd also propose to test with more recent PHP version.

**Comment by ghmail** [comment]
My steps were as follows:
1. Open the message editor to write anew mail.
2. Switch to "HTML" (if not already set via user setting. Doesn't matter).
3. Save mail (got the confirmation that save was successful).
4. Go back to folders (leave editor).
5. Open Draft folder and try to open message for editing.
6. Because of wrong Content-Type (text/plain) I see only the raw text in the editor.

This is reproducible with any skin.

Regarding newer PHP: I'm bound to the version of my distro (7.2.5). As long as I have no indication that a wrong header value is caused by my particular PHP version, I cannot migrate my server installation to a external PHP-repo.

There is NO system-wide `mimePart.php` installed.

As I mentioned above: I think it should be possible to isolate the code that is responsible to set the Content-Type value in case of HTML and why this code is making a wrong decision under certain circumstances (given by the systems PHP installation or whatever). I would say some extra log output could help.

Maybe there could also be a way for remote debugging via xdebug.

**Comment by cmichel44** [comment]
Hi @ghmail and @alecpl,

I do have the same problem with incorrect Content-Type on all of my outgoing emails. Interestingly I am also running on openSUSE 15.1.

I only realized this behavior recently and first thought it was caused by the update to roundcube 1.4.4 because I saw correct header on old messages created with 1.4.4. But it may be that the roundcube update also coincided with package upgrades on the distribution.

If there is any way to debug this or adding additional log statements to code, I am happy to help. Meanwhile I will search if I can find what packages were updated recently.

Thanks, Christian

**Comment by ghmail** [comment]
It is an issue with the current mail_mime (1.10.8) bundled with Roundcube.

I just replaced the files `mime.php` and `mimePart.php` with files from an older Roundcube package (version 1.3.6) in _roundcube/vendor/pear/mail_mime/Mail/_.

Now everything with HTML works fine!

As a conclusion I would say the problem is somewhere in newer versions of mail_mime.

**Comment by alecpl** [comment]
Problem is it works for me with 1.10.8.

**Comment by cmichel44** [comment]
I just did a quick test, 1.10.5 still fails, 1.10.4 gives me a good result.

It's strange that Alec does not see the problem with the current version though.

**Comment by ghmail** [comment]
Confirmative. I also downgraded to 1.10.4 and it works.

Maybe the mail_mime developers could say what they have changed between 1.10.4 and 1.10.5 regarding PHP 7.2.5 (I think @alecpl is using a newer version).

**Comment by cmichel44** [comment]
I am wondering whether this part in method header() is causing trouble, where the Content-Type is removed unconditionally:

        // Always reset Content-Type/Content-Transfer-Encoding headers
        // In case the message structure changed in meantime
        unset($this->headers['Content-Type']);

And maybe we're not in a condition afterwards that will regenerate this header field. I can look at this more tonight, still have some work to do ;-)

**Comment by alecpl** [comment]
BTW, I'm the "Mail_mime developers" ;)

**Comment by alecpl** [comment]
Maybe you could clone https://github.com/pear/Mail_Mime and run the tests on your system?

**Comment by cmichel44** [comment]
I will use the github source as a base for my testing.

**Comment by ghmail** [comment]
I have a dev machine with openSUSE 15.0 but with also with PHP 7.2.5.

The unittest run fine.
[Mail_Mime-phpunit.log.xml.txt](https://github.com/roundcube/roundcubemail/files/4787222/Mail_Mime-phpunit.log.xml.txt)

**Comment by alecpl** [comment]
Let's focus on a save-draft action, as sending email might hit a different code path. This change may give some hints:
```diff
--- a/program/include/rcmail_sendmail.php
+++ b/program/include/rcmail_sendmail.php
@@ -492,8 +492,10 @@ class rcmail_sendmail
                     }
                 }
                 else {
+rcube::console($message);
                     $msg     = $message->getMessage();
                     $headers = '';
+rcube::console($msg);
                 }
 
                 if (is_a($msg, 'PEAR_Error')) {

```

**Comment by ghmail** [comment]
with mail_mime version 1.10.8 and your patch applied:

[console.log.txt](https://github.com/roundcube/roundcubemail/files/4788203/console.log.txt)

**Comment by cmichel44** [comment]
Alec, I am looking at a very crazy behaviour. It looks like PHP is going crazy in mime.php  contentHeaders(). If I simply change the return on the default clause to a break, it works again. But it does not make sense. I also added code to dump all the boolean flags, which made perfect sense and should direct the code to the multipart/alternative path, and it actually went there. I could even reinstate the original default path and it still worked. When I remove my debug code to write the variables, it fails again.

To me that looks like PHP is having an uninitialized memory issue, but I cannot really proof it. Also the last PHP maintenance is longer ago than when I realized the problem. Very confusing.

**Comment by alecpl** [comment]
So, for you it is:
```
'Content-Type' => 'text/plain; charset=US-ASCII;
 format=flowed',
```
for it is:
```
'Content-Type' => 'multipart/alternative;
 boundary="=_db8ca0715e7525d369c99c0fbbf7cdae"',
```

Do you have opcache enabled? Try to disable it.

**Comment by cmichel44** [comment]
opcache is the culprit, as soon as I disable it, I get the header formatted correctly :-(.
Well at least now I know where it comes from, not why. Let's see if @ghmail can confirm this as well.

Thanks a lot for your support, Alec!

**Comment by ghmail** [comment]
Yes, I can confirm: disabling opcache in my _php.ini_ solved the HTML issue. Tested with mail_mime version 1.10.8.

Thanks from me too, Alec!

**Comment by alecpl** [comment]
I guess it would be good to create a ticket at OpenSuse.

@cmichel44, if you could spend some more time on it, could you check if replacing the problematic `switch` statement with an `if` would fix the issue?

ps. I cannot reproduce the issue with opcache enabled (with default opcache config) and PHP 7.2.31.

**Comment by cmichel44** [comment]
Replacing the switch with cascaded ifs didn't help. What actually helps is removing the default and after the switch block adding:
if (empty($headers))
            return $headers;

I have to check how I can open a ticket for OpenSuse and how to describe it.

**Comment by cmichel44** [comment]
No need to open a bug, they are kind of aware of this problem since March:
https://bugzilla.opensuse.org/show_bug.cgi?id=1166235

Interesting: they also tried moving the creation of the $headers array to the top and that also seems to fix the problem, and that works for me as well.

So, I know that it's not a problem in the mime code, but a reorder of the coding to move the creation of the $headers array to the beginning actually circumvents the openSUSE problem. Not sure if you want to add such a workaround to your code.

**Comment by ghmail** [comment]
If a enabled PHP opcache results in unpredictable behaviour of a PHP application, then this is a really serious security issue, isn't it?

**Comment by cmichel44** [comment]
Thanks Alec for adding this workaround. I hope that this will help other users of openSUSE as well. Let me know when you pushed to master and I will retry once again with the original file from github, just to make sure ;-)

**Comment by alecpl** [comment]
@cmichel44 it has been pushed to master. If you confirm the fix, I'll create a new package release.

**Comment by cmichel44** [comment]
@alecpl Sorry for the delay. I am buried in deep trouble in my real job since this afternoon, so only saw your message right now. I have cloned git and re-ran my test and I can confirm that it works while I have opcache enabled. Thanks for your support to get this resolved. Now my mails will stop looking so crazy ;-).

**Comment by ghmail** [comment]
As the orginator of this issue I also can confirm that your fix works fine on openSUSE 15.1 with opcache enabled.

Many thanks for your great support.

**Comment by alecpl** [comment]
Mail_mime-1.10.9 has been released.
